text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,178 @@
+configure resource information.
+
+Note there are some existing implementations of this already in the SDKs (see
+[below](#prior-art-and-alternatives)), but nothing currently in the
+specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to all kinds of telemetry,
+a user will need to configure which resource detector(s) they would like to run
+(e.g. AWS EC2 detector).
+
+If multiple detectors are configured, and more than one of these successfully
+detects a resource, the resources will be merged according to the Merge
+interface already defined in the specification, i.e. the earliest matched
+resource's attributes will take precedence. Each detector may be run in
+parallel, but to ensure deterministic results, the resources must be merged in
+the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the
+`OTEL_RESOURCE` environment variable will be included in the SDK. The
+environment variable will contain of a list of key value pairs, and these are
+expected to be represented in a format similar to the [W3C
+Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value),
+except that additional semi-colon delimited metadata is not supported, i.e.:
+`key1=value1,key2=value2`. If the user does not specify any resource, this
+detector will be run by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud
+vendors) must be implemented as packages separate to the core SDK, and users
+will need to import these separately.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK
+specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for a global function to merge resources returned by a set of
+  detectors
+- Details of the "from environment variable" detector implementation as
+  described above
+- Specification that default detection (from environment variable) runs once on
+  startup, and is used by all tracer & meter providers by default if no custom
+  resource is supplied
+
+### Usage
+
+The following example in Go creates a tracer and meter provider that uses
+resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel/api`, `otel/sdk`,
+`otel/awsdetector`, and `otel/gcpdetector` packages.
+
+```go
+resource, _ := sdkresource.Detect(ctx, 5 * time.Second, awsdetector.ec2, gcpdetector.gce)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+mp := push.New(..., push.WithResource(resource))
+```
+
+### Components
+
+#### Detector
+
+The `Detector` interface will simply contain a `Detect` function that returns a
+Resource.
+
+The `Detect` function should contain a mechanism to timeout and cancel the
+request. If a detector is not able to detect a resource, it must return an
```

Comment:
> the request

Which request? First time the word "request" appears in this OTEP. I think resource detection must not perform network I/O, as resources are immutable and thus initializing resources must happen before telemetry can be collected. A timeout is out of the question.

One way to lift this restriction would be to specify a way to do this asynchronously, and delay the first call to export() until resources are detected. But this warrants a whole new OTEP. For now, I'd say anything which could require a timeout is out of the question as input for resources.

Reply:
I'm not sure that this is something that needs to be in the spec, but at least in the Go SDK this timeout/cancel mechanism would be provided by the `context` referenced in [this comment](https://github.com/open-telemetry/oteps/pull/111#discussion_r448612875).  I think that's why it would be important to include the `context` even if it were not being used as an OpenTelemetry Context.

Reply:
> Which request? First time the word "request" appears in this OTEP. I think resource detection must not perform network I/O, as resources are immutable and thus initializing resources must happen before telemetry can be collected.

Changed the wording to `operation` since not all detectors would not be creating any kind of request. I don't think it's unreasonable for detectors to perform network I/O synchronously if it is expected to be (relatively) quick, especially if a user specifically pulls in that package. Indeed I think would not be uncommon for looking up resource information: I think the convention by all major cloud vendors is to provide instance metadata via `http://169.254.169.254`

> One way to lift this restriction would be to specify a way to do this asynchronously, and delay the first call to export() until resources are detected. But this warrants a whole new OTEP. For now, I'd say anything which could require a timeout is out of the question as input for resources.

I don't think this would be worth the complexity. We'd have to make sure the resource information is applied to any "pending" telemetry before being exported which would not be trivial.

Reply:
Good point with it being the users choice. I just hope that most resources can be provided via things like environment variables or local files. For example, cloud instance metadata could probably queried by the backend if the resource contains some sort of instance ID?